File: sdks/browser4-cli/src/cli/daemon/daemon.ts
/**
 * Copyright (c) Microsoft Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/*
 * This file implements the daemon server that listens on a socket.
 * It handles client connections, processes incoming commands, and manages
 * the lifecycle of the browser context and backend server.
 */

import * as fs from 'fs';
import path from 'path';

import * as os from 'os';
import {spawn} from 'child_process';
import * as https from 'https';

import {commands} from './commands';
import {parseCommand} from './command';
import axios, {AxiosInstance} from "axios";
import {readState} from "../../state";

import type * as mcp from './../../mcp/exports';

function makeAxios(baseUrl: string): AxiosInstance {
    return axios.create({
        baseURL: baseUrl.replace(/\/$/, ''),
        timeout: 30_000,
        headers: {'Content-Type': 'application/json'},
    });
}

export async function ensureServerRunning(args: string[]): Promise<void> {
    const state = readState();
    let baseUrl = state.baseUrl || 'http://localhost:8182';

    // Check for --server override in args
    const serverIdx = args.indexOf('--server');
    if (serverIdx !== -1 && args[serverIdx + 1]) {
        baseUrl = args[serverIdx + 1];
    }

    // Only attempt to start if we're pointing to localhost
    if (!baseUrl.includes('localhost') && !baseUrl.includes('127.0.0.1')) {
        return;
    }

    // Check if already running
    try {
        const ax = makeAxios(baseUrl);
        await ax.get('/actuator/health');
        return;
    } catch (error) {
        // Not running, proceed to start
    }

    console.log('Browser4 server not running. Starting...');

    const jarPath = await findOrDownloadJar();
    const port = parseInt(new URL(baseUrl).port) || 8182;

    await startServer(jarPath, port);
}

async function findOrDownloadJar(): Promise<string> {
    // Check environment variable
    if (process.env.BROWSER4_JAR_PATH && fs.existsSync(process.env.BROWSER4_JAR_PATH)) {
        return process.env.BROWSER4_JAR_PATH;
    }

    // Check common locations
    const candidates = [
        path.join(os.homedir(), '.browser4', 'Browser4.jar'),
        path.resolve('Browser4.jar'),
        path.resolve('target', 'Browser4.jar'),
        path.resolve(__dirname, '..', '..', 'target', 'Browser4.jar'),
        path.resolve(__dirname, '..', '..', '..', 'browser4', 'browser4-agents', 'target', 'Browser4.jar'), // Monorepo location
    ];

    for (const candidate of candidates) {
        if (fs.existsSync(candidate)) {
            return candidate;
        }
    }

    // Download if not found
    const downloadPath = path.join(os.homedir(), '.browser4', 'Browser4.jar');
    await downloadJar(downloadPath);
    return downloadPath;
}

async function downloadJar(targetPath: string): Promise<void> {
    const dir = path.dirname(targetPath);
    if (!fs.existsSync(dir)) {
        fs.mkdirSync(dir, {recursive: true});
    }

    const url = 'https://github.com/platonai/Browser4/releases/latest/download/Browser4.jar';
    console.log(`Downloading Browser4.jar from ${url}...`);

    return new Promise((resolve, reject) => {
        const file = fs.createWriteStream(targetPath);
        https.get(url, (response) => {
            if (response.statusCode === 301 || response.statusCode === 302) {
                if (!response.headers.location) {
                    reject(new Error('Redirect without location header'));
                    return;
                }
                https.get(response.headers.location, (redirectResponse) => {
                    redirectResponse.pipe(file);
                    file.on('finish', () => {
                        file.close();
                        console.log('Download complete');
                        resolve();
                    });
                }).on('error', (err) => {
                    fs.unlinkSync(targetPath);
                    reject(err);
                });
            } else {
                response.pipe(file);
                file.on('finish', () => {
                    file.close();
                    console.log('Download complete');
                    resolve();
                });
            }
        }).on('error', (err) => {
            fs.unlinkSync(targetPath);
            reject(err);
        });
    });
}

async function startServer(jarPath: string, port: number): Promise<void> {
    console.log(`Starting server from ${jarPath} on port ${port}...`);

    const child = spawn('java', ['-jar', jarPath, `--server.port=${port}`], {
        detached: true,
        stdio: 'ignore' // or 'inherit' for debugging, but 'ignore' keeps it clean
    });

    child.unref(); // Allow the parent process to exit independently

    // Wait for health check
    const start = Date.now();
    const timeout = 60000;

    while (Date.now() - start < timeout) {
        try {
            await axios.get(`http://localhost:${port}/actuator/health`);
            console.log('Server is up and running.');
            return;
        } catch (e) {
            await new Promise(resolve => setTimeout(resolve, 1000));
        }
    }

    throw new Error(`Server failed to start within ${timeout}ms`);
}

function formatResult(result: mcp.CallToolResult) {
    // Formats the MCP tool execution result for the client response.
    const isError = result.isError;
    const text = result.content[0].type === 'text' ? result.content[0].text : undefined;
    return {isError, text};
}

function parseCliCommand(args: Record<string, string> & { _: string[] }): {
    toolName: string,
    toolParams: NonNullable<mcp.CallToolRequest['params']['arguments']>
} {
    // Parses CLI arguments to identify the command and its parameters.
    const command = commands[args._[0]];
    if (!command)
        throw new Error('Command is required');
    return parseCommand(command, args);
}

File: sdks/browser4-cli-rust/src/daemon.rs
//! Server daemon management for the Browser4 CLI.
//!
//! Ensures a Browser4 server is running before executing commands.
//! Only manages localhost instances; remote servers are not touched.

use std::fs;
use std::path::PathBuf;
use std::process::{Command, Stdio};

use reqwest::Client;

use crate::managed_processes::{register_managed_server_process, ManagedServerProcess};
use crate::state::read_state;

/// Ensure the Browser4 server is running, starting it if necessary.
///
/// Only acts on `localhost` / `127.0.0.1` URLs.
pub async fn ensure_server_running(base_url: &str) -> Result<(), String> {
    // Skip remote servers
    if !base_url.contains("localhost") && !base_url.contains("127.0.0.1") {
        return Ok(());
    }

    // Check if already running
    let client = Client::builder()
        .timeout(std::time::Duration::from_secs(5))
        .build()
        .map_err(|e| e.to_string())?;

    let health_url = format!("{}/actuator/health", base_url.trim_end_matches('/'));
    if client.get(&health_url).send().await.is_ok() {
        return Ok(());
    }

    eprintln!("Browser4 server not running. Starting...");

    let jar_path = find_or_download_jar().await?;
    eprintln!("Jar path: {}", jar_path.display());

    let port = extract_port(base_url);
    start_server(&jar_path, base_url, port).await
}

fn extract_port(base_url: &str) -> u16 {
    if let Ok(url) = reqwest::Url::parse(base_url) {
        url.port().unwrap_or(8182)
    } else {
        8182
    }
}

async fn find_or_download_jar() -> Result<PathBuf, String> {
    // Check environment variable
    if let Ok(env_path) = std::env::var("BROWSER4_JAR_PATH") {
        let p = PathBuf::from(&env_path);
        if p.exists() {
            return Ok(p);
        }
    }

    // Common candidate locations
    let home = dirs::home_dir().unwrap_or_else(|| PathBuf::from("."));
    let candidates = vec![
        PathBuf::from("Browser4.jar"),
        PathBuf::from("target/Browser4.jar"),
        home.join(".browser4").join("Browser4.jar"),
        home.join(".browser4").join("lib").join("Browser4.jar"),
    ];

    for candidate in candidates {
        if candidate.exists() {
            return Ok(candidate);
        }
    }

    // Download if not found
    let download_path = home.join(".browser4").join("lib").join("Browser4.jar");
    download_jar(&download_path).await?;
    Ok(download_path)
}

async fn download_jar(target_path: &PathBuf) -> Result<(), String> {
    if let Some(dir) = target_path.parent() {
        fs::create_dir_all(dir).map_err(|e| e.to_string())?;
    }

    let url = "https://github.com/platonai/Browser4/releases/latest/download/Browser4.jar";
    eprintln!("Downloading Browser4.jar from {}...", url);

    let client = Client::builder()
        .timeout(std::time::Duration::from_secs(300))
        .build()
        .map_err(|e| e.to_string())?;

    let response = client
        .get(url)
        .send()
        .await
        .map_err(|e| format!("Download failed: {e}"))?;

    if !response.status().is_success() {
        return Err(format!(
            "Download failed with status: {}",
            response.status()
        ));
    }

    let bytes = response.bytes().await.map_err(|e| e.to_string())?;
    fs::write(target_path, &bytes).map_err(|e| e.to_string())?;

    eprintln!("Download complete.");
    Ok(())
}

async fn start_server(jar_path: &PathBuf, base_url: &str, port: u16) -> Result<(), String> {
    eprintln!(
        "Starting server from {} on port {}...",
        jar_path.display(),
        port
    );

    let child = Command::new("java")
        .args([
            "-jar",
            jar_path.to_str().unwrap_or("Browser4.jar"),
            &format!("--server.port={}", port),
        ])
        .stdin(Stdio::null())
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .spawn()
        .map_err(|e| format!("Failed to start server: {e}"))?;

    let pid = child.id();

    register_managed_server_process(
        ManagedServerProcess {
            pid,
            base_url: base_url.to_string(),
            port,
            jar_path: jar_path.to_string_lossy().to_string(),
            started_at: chrono::Utc::now().to_rfc3339(),
        },
        None,
    );

    // Detach: we drop the Child handle here. The spawned process continues
    // running independently because we set all stdio to null and call drop().
    // On Unix this means the child is still a child of this process until we
    // exit, at which point it is re-parented to init/systemd. That is
    // sufficient for our use-case where the parent CLI exits immediately.
    drop(child);

    // Wait for health check (up to 60 seconds)
    let client = Client::builder()
        .timeout(std::time::Duration::from_secs(5))
        .build()
        .map_err(|e| e.to_string())?;

    let health_url = format!("http://localhost:{}/actuator/health", port);
    let start = std::time::Instant::now();
    let timeout = std::time::Duration::from_secs(60);

    loop {
        tokio::time::sleep(std::time::Duration::from_secs(1)).await;
        if client.get(&health_url).send().await.is_ok() {
            eprintln!("Server is up and running.");
            return Ok(());
        }
        if start.elapsed() > timeout {
            return Err(format!(
                "Server failed to start within {}s",
                timeout.as_secs()
            ));
        }
    }
}

/// Resolve the base URL from CLI state + optional server override arg.
pub fn resolve_base_url(override_url: Option<&str>) -> String {
    let state = read_state(None);
    let base = override_url
        .map(|s| s.to_string())
        .unwrap_or(state.base_url);
    base.trim_end_matches('/').to_string()
}
